### PR TITLE
Adding some CTS variants for indirect command buffers.

### PR DESCRIPTION
**Pull request by benvanik**
This is a simple start that tests a few commands to test the infrastructure for record/replay of indirect command buffers and the validation mechanism. We don't have a good way in the CTS yet to conditionally run tests based on whether validation is compiled into the build but I'll be looking into that for testing failure cases in future PRs (for now I've just tested manually to verify errors are propagated).

### CODE REVIEW

**Comment by benvanik** [comment]
lol vmvx tsan failure is legit - that's cool - no other tsan tests failed with what is definitely an incorrect test input

**Review by ScottTodd — APPROVED** [review]
Nice!

**Comment by benvanik** [comment]
cool, one last leak to fix <3 asan

**Comment by benvanik** [comment]
ok, darn, that nvidia vulkan flake is happening quite a bit now
```
[ RUN      ] CommandBufferTest/CommandBufferFillTest.FillBuffer_pattern2_size16_offset0_length10/indirect
/work/runtime/src/iree/hal/cts/cts_test_base.h:86: Failure
Value of: status
Expected: error code OK
  Actual: 0x16a8449, whose error code is FAILED_PRECONDITION: /work/runtime/src/iree/hal/drivers/vulkan/vulkan_driver.cc:251: FAILED_PRECONDITION; VK_ERROR_INCOMPATIBLE_DRIVER; vkCreateInstance: invalid instance configuration
<and fails forevermore>
```

I think this is just because of the increased number of tests that are cycling through devices.

**Comment by benvanik** [comment]
confirmed it's the count (most likely) - it failed twice in a row on `CommandBufferTest/CommandBufferFillTest.FillBuffer_pattern2_size16_offset0_length10/indirect` so I disabled all the direct (existing codepath tests) such that it only runs `indirect` and now it passes.

I'll take a look at caching devices as part of this.

**Comment by benvanik** [comment]
I split out some tests into different files (they were getting hard to manage in a single one) and made them reuse devices 🤞

File: runtime/src/iree/hal/cts/driver_test.h [hunk 1]
```diff
@@ -16,19 +16,44 @@
...
 
-namespace iree {
-namespace hal {
-namespace cts {
+namespace iree::hal::cts {
 
-class driver_test : public CTSTestBase<> {
+// NOTE: does not use CTSTestBase as we don't want the automatically created
+// driver/device it provides.
+class DriverTest : public ::testing::Test {
```
Comment (ScottTodd) [inline runtime/src/iree/hal/cts/driver_test.h:23]: Good idea - lets the other tests reuse devices while this test works as it did before.

**Review by ScottTodd — APPROVED** [review]

**Pull request merged**